### ISSUE TITLE
Exponential notation (e.g., 1e+140) is not parsed as a BigInt but as a double

### ISSUE THREAD

**Issue by thumdugger** [ozaner/json_bigint#1]
Hi,

While trying to use `json_bigint` for learning about writing **Flutter** and **dart** apps I discovered the package doesn't recognize exponential notation for integers.

Using the example code as illustration,

```
String json =
    '{"sometext": "hello world!", "bignumber": 100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000}';
```

correctly parses *bignumber* as a `BigInt`, but

```String json = '{"sometext": "hello world!", "bignumber": 1e+140}';```

incorrectly parses *bignumber* as a `double` (just as convert does).

In my current use case I am pulling the number from an external website so have no direct control over when it decides to hand me an integer in integer notation or in exponential notation. 

(Also, as a side note, the example code is missing the '}' closing character on the *json* declaration such that running the example code results in a runtime parsing error.)

**Comment by thumdugger** [comment]
So I'm thinking the issue is caused by Dart itself.

At a low level Dart appears to convert numbers like 1e+140 to doubles, automatically, with no apparent option to keep it from doing so. Worse, even trying to sidestep the problem by using `BigInt.parse('1e+140')` results in a FormatException as, apparently, BigInt, itself, doesn't handle exponential notation, at all.

I was originally fooled by the documentation on this because I saw that `BigInt.parse('7e5', radix: 16)` worked fine until I realized it worked fine because `e` is a hex value not an exponential indicator.

Not sure why `BigInt` was designed in such a strangely broken way but at least I understand what it is doing now. I'm too new to Dart to know how this is supposed to be handled but I can only assume most, if not all, pathways will end up with big exponential notation values getting converted to a double which, basically, just sort of breaks everything.

Hopefully there is an answer out there but the only answer, at the moment, that I can see here is if we altered the parsing code to convert things like 1e+140 to a canonical string, first, and then parsing that with `BigInt.parse()` from there. The bigger problem would be going the other way as I'm thinking there are probably many edge cases that would get in the way such that `encodeJson(decodeJson('1e+140'))` may not be `1e+140`.

**Comment by ozaner** [comment]
I think you're right about converting it to a canonical form. I'm gonna try doing that.
As for the errors in the example code, I've fixed that, thanks for point it out.

**Comment by thumdugger** [comment]
I played with it a little a couple days ago as a _Dart_ learning exercise. The `numberPrimitive` `petitparser` matcher already captures the form correctly so I thought about separating it into an integer and exponential notation form (which might be a decent answer) and handling it there but instead started modifying the `_numOrBigInt` method.

Immediately ran into the first "interesting" design problems: 

1. to what extent should a BigInt raised to the BigInt be supported?
2. should an arbitrary chaining of exponents be supported? (.i.e, X1^Y2...^Xn)?
3. should there be an upper limit to an allowed value?
4. would it just be easier to have an option to decode exponential notation numbers (or any numbers, actually) into strings instead of turning them into number?

 The third one struck me as important lest some malicious, or prankster, number source present a googolplex raised to the googolplex (10^10^100)^(10^10^100) raised to the ... you get the idea. Since "arbitrary" would let someone do such things, especially if chaining was supported, it would a lame if there was no way to control or limit that behavior.

While pondering how one might handle such issues it occurred to me that the allowing BigInt values to just not be turned into numbers might also be an answer. The tutorial I was going through only used the values for presenting (which would work fine), not calculating, and, more importantly, the website that the code interacts with uses the numbers for returning random trivia about the number which led to this additional, possible, feature for the encodeJson encoder:

5. allow strings formatted as numbers to be encoded as numbers instead of strings.

This last part would help support the canonical and reversible portions of the parser.

I started down this path, wrote some tests, updated the example code (to also include using the EncoderSettings variable which the original defined but never used), and then went "but what does ozaner want to do? Maybe the vision is entirely different than anything I just laid out." 

So I sort of just stopped until maybe I heard back or just became super interested in seeing some of these changes I was pondering actually work. I did fork it off to safely play with it so if there was anything you found interesting (really only the example code is current in the forked repo) then feel free to grab it or let me know and maybe I could learn how to submit patches on github and send it your way all official like and stuff).

Thanks for making and posting the code. It has been fun to play with and learn from. And it also introduced me to petitparser which, it just so happens, is exactly the library I was needing for another little project piece of Dart code I'm playing around with right now. So `json_bigint` has been a big hit for me!

**Comment by ozaner** [comment]
Ok I released an updated version of the package as v2.0.0, that adds support for this. You can see the documentation on [pub.dev](https://pub.dev/packages/json_bigint), but what's relevant to this discussion is that it adds `DecoderSettings`.

By default, the decoder still treats any number written in exponential notation (e.g. 1.212e+3, 2e4, 3e-4, etc.) as a double. However, when  `treatExpAsIntWhenPossible` (one of the new decoder settings) is set to true, it tries to parse these exponential numbers and determine if they would evaluate to integers. If so, it converts them to integers instead of doubles.

For example:
- 1e6 = 1000000 is an int so it becomes a BigInt.
- -23000e-3 = -23 is an int so it becomes a BigInt.
- 2.312e2 = 231.2 is not an int so it becomes a double.
- 43e-2 = .42 is not an int so it becomes a double.

You can look at the source code if you're interested, but basically I just made a [custom regex expression](../blob/main/lib/src/decoder.dart#L190) that runs on the string should it fail to parse as a BigInt. Then I do some calculations to make sure it evaluates to an integer with no decimal part, and return it as a BigInt if so.

As for representing large numbers as strings, this is possible and is the most common way of getting around the problem of decoding BigIntegers from JSON in a lot of languages. You can just encode integers as strings and convert them back to integers yourself.

For my use case though, I didn't want to have my output JSON include quotations around the numbers. This meant I needed to be able to truly serialize BigInts rather than just their string counterparts. This is ultimately is why I made this library.

Anyway thanks for using and experimenting with the library, I hope this update allows you to decode exponentials as BigInts. Assuming I didn't break anything. If it doesn't work as intended, you can just reopen the issue.
